olive/passes/onnx/mnb_to_qdq.py

```diff
@@ -62,257 +64,125 @@ def _run_for_config(
     ) -> ONNXModelHandler:
         output_model_path = resolve_onnx_path(output_model_path, Path(model.model_path).name)
 
-        # create a dag from the model
-        dag = OnnxDAG.from_model_path(model.model_path)
-        # remove unnecessary identity nodes
-        dag.remove_identity_nodes()
-
-        # if matmulnbits zero point is the following, then the zero point is not needed in the DQ node
-        default_mnb_zp = 8 if config["use_int4"] else 0
-        int_np_dtype = np.int8 if config["use_int4"] else np.uint8
-        int_elem_type = onnx.TensorProto.INT4 if config["use_int4"] else onnx.TensorProto.UINT4
-
-        num_modified = 0
-        for node_name in dag.get_node_names():
-            op_type = dag.get_node_op_type(node_name)
-            if op_type != "MatMulNBits":
-                continue
-
-            node_inputs = dag.get_node_inputs(node_name)
-            # only deal with the constant matmul case for now
-            if not all(dag.is_initializer(i_name) and not dag.is_input(i_name) for i_name in node_inputs[1:]):
-                continue
-
-            graph_idx = dag.get_graph_idx(node_name)
+        # 2 Step
+        # 1. pattern replacement
+        # 2. Repacking
+        ir_model = ir.serde.deserialize_model(model.load_model())
+
+        def mat_mul_n_bits_pattern(op, input_A, q_weight, q_scales, q_zeros, g_idx, bias):
+            # bias is an optional input
+            return op.MatMulNBits(
+                input_A,
+                q_weight,
+                q_scales,
+                q_zeros,
+                g_idx,
+                bias,
+                _outputs=["mat_mul_n_bits_out"],  # Bind the output to the name "mat_mul_n_bits_out"
+            )
 
-            # original output proto
-            node_output = dag.get_node_outputs(node_name)[0]
-            is_model_output = dag.is_output(node_output)
-            node_output_proto = None
-            if dag.get_io(node_output).proto:
-                node_output_proto = dag.get_io(node_output).proto[-1]
-            node_attributes = dag.get_node_attributes(node_name)
-            K = node_attributes["K"]  # noqa: N806
-            N = node_attributes["N"]  # noqa: N806
-            block_size = node_attributes["block_size"]
-            num_k_blocks = math.ceil(K / block_size)
+        def _is_initializer(context, value: ir.Value) -> bool:
+            graph: ir.Graph = context.graph
+            return value in graph.initializers.values()
+
+        def mat_mul_n_bits_pattern_check(context, *, q_weight, g_idx, mat_mul_n_bits_out: ir.Value, **_) -> bool:
```

Comment:
does q_weight here match for the input right before g_idx or it is whatever it is in the mat_mul_n_bits_pattern signature? The input begore g_idx is qzero and can be optional.  we want to match the second input,

Reply:
@gramalingam

Reply:
The inputs of the pattern-function (mat_mul_n_bits_pattern) are bound to values in the graph, and these values are passed in as keyword-arguments to the rewrite function here. So, the order here doesn't really matter, though I usually just copy-paste and use the same argument list for both.